core/src/test/kotlin/org/eln2/sim/electrical/mna/CircuitTest.kt

```diff
@@ -70,6 +93,86 @@ internal class CircuitTest {
 			assertEquals(ts.vs.potential, ts.r1.potential, EPSILON)
 		}
 	}
+	
+	@Test
+	fun componentChangeConsistency() { // Also tests resistors in parallel.
+		val ts = TrivialResistiveCircuit()
+		ts.c.step(0.5)
+		val current = ts.r1.current
+
+		val r2 = Resistor()
+		ts.c.add(r2)
+		r2.resistance = ts.r1.resistance
+		r2.connect(0, ts.r1, 0)
+		r2.connect(1, ts.r1, 1)
+
+		ts.c.step(0.5)
+		assertEquals(r2.current, ts.r1.current, EPSILON)
+		assertEquals(ts.r1.current, current, EPSILON) {"i1_1=${ts.r1.current} i1_0=$current i2=${r2.current}"}
+		assertEquals(ts.vs.current, current * 2.0, EPSILON)
+
+		ts.c.remove(r2)
+
+		ts.c.step(0.5)
+		assertEquals(ts.r1.current, current, EPSILON)
+		assertEquals(r2.circuit, null)
+	}
+
+	@Test
+	fun resistorsInSeries() {
+		val circuit = Circuit()
+		val vs = VoltageSource()
+		val r1 = Resistor()
+		val r2 = Resistor()
+
+		vs.potential = 10.0
+		r1.resistance = 5.0
+		r2.resistance = 5.0
+
+		circuit.add(vs, r1, r2)
+		vs.connect(1, r1, 1)
+		r1.connect(0, r2, 1)
+		r2.connect(0, vs, 0)
+		vs.connect(1, circuit.ground)
+
+		circuit.step(0.5)
+		assertEquals(r1.current, r2.current, EPSILON)
+		assertEquals(r1.current, 1.0, EPSILON)
+	}
+
+	@Test
+	fun basicCapacitorCircuit() {
+		val circuit = Circuit()
+		val vs = VoltageSource()
+		val c1 = Capacitor()
+
+		vs.potential = 10.0
+		c1.capacitance = 0.01
+
+		circuit.add(vs, c1)
+		vs.connect(1, c1, 1)
+		vs.connect(0, c1, 0)
+		vs.connect(1, circuit.ground)
+
+		circuit.step(0.5)
+		circuit.step(0.5)
+		circuit.step(0.5)
+		circuit.step(0.5)
+	}
+
+	@Test
+	fun basicRCCircuit() {
+		val ts = TrivialRCCircuit()
+		ts.c.step(0.05)
+		assertEquals(ts.c1.current, 0.000732, 0.000001)
+		assertEquals(ts.c1.potential, 6.315, 0.001)
+		ts.c.step(0.05)
+		assertEquals(ts.c1.current, 0.000270, 0.000001)
+		assertEquals(ts.c1.potential, 8.660, 0.001)
+		ts.c.step(0.05)
+		assertEquals(ts.c1.current, 0.000099, 0.000001)
+		assertEquals(ts.c1.potential, 9.506, 0.001)
+	}
```

Comment:
@Grissess these tests are backwards... it is `expected`, `actual`, `interval`.

Reply:
That argument order annoys me--I'd much rather rebind the function swapping arguments than rewrite those tests. (That's how you read equalities like this: "assert [value under test] equals [compared value"--unless you happen to be one of those "defensive" programmers that deliberately avoids putting an lvalue on the left hand side so as to accidentally catch "foo = 7" assignment expression bugs.)

The function doesn't change in either case :)

Reply:
No, the function doesn't change, but the debug printouts do. If the order doesn't functionally change anything, I'd much rather have the proper assertions using the code as designed, not as convenient, that way it debug prints properly.

Reply:
Which reminds me - in my tests I handled it differently - I handled everything as a boolean statement, and then I just did my own statement in there and made sure that if things were good that it passed the assertion. This had the added benefit of allowing me to debug this in the testing logic, rather than dying in the test assertions and not being able to output debugging information. Arguably here, the best solution would be to change how those tests are run so that if it plans to fail, that it instead exports the graph of what it was doing and *then* it calls in the assertion and fails. That allows us to have better control over our debug logic and include debugging output as part of the testing infrastructure.

Reply:
Cael suggested this may be a bit over-the-top, so changing the tests to use the methods correctly would be sufficient I think.